visualization/app/codeCharta/ui/dialog/dialog.globalSettings.component.spec.ts

```diff
@@ -0,0 +1,106 @@
+import { DEFAULT_SETTINGS } from "../../util/dataMocks"
+import { IRootScopeService } from "angular"
+import _ from "lodash"
+
+describe("DialogGlobalSettingsController", () => {
+	let dialogGlobalSettingsController: DialogGlobalSettingsController
+	let $mdDialog
+	let settingsService: SettingsService
+	let $rootScope: IRootScopeService
+	let settings: Settings
+
+	beforeEach(() => {
+		restartSystem()
+		withMockedSettingsService()
+		rebuildController()
+	})
+
+	function restartSystem() {
+		instantiateModule("app.codeCharta.ui.dialog")
+
+		$mdDialog = getService("$mdDialog")
+		settingsService = getService<SettingsService>("settingsService")
+		$rootScope = getService<IRootScopeService>("$rootScope")
+
+		settings = _.cloneDeep(DEFAULT_SETTINGS)
+	}
+
+	function rebuildController() {
+		dialogGlobalSettingsController = new DialogGlobalSettingsController($mdDialog, $rootScope, settingsService)
+	}
+
+	function withMockedSettingsService(mockSettings: Settings = settings) {
+		settingsService = jest.fn<SettingsService>(() => {
+			return {
+				getSettings: jest.fn().mockReturnValue(mockSettings)
+			}
+		})()
+	}
+
+	describe("constructor", () => {
+		it("should subscribe to SettingsService Events", () => {
+			SettingsService.subscribe = jest.fn()
+
+			rebuildController()
+
+			expect(SettingsService.subscribe).toHaveBeenCalledWith($rootScope, dialogGlobalSettingsController)
+		})
+	})
+
+	describe("updateSettingsFields", () => {
+		it("should update viewModel.hideFlatBuildings with appSettings", () => {
+			dialogGlobalSettingsController["_viewModel"] = { hideFlatBuildings: null } as any
+			settings.appSettings.hideFlatBuildings = false
+
+			dialogGlobalSettingsController.updateSettingsFields(settings)
+
+			expect(dialogGlobalSettingsController["_viewModel"].hideFlatBuildings).toBeFalsy()
+		})
+
+		it("should update viewModel.isWhiteBackground with appSettings", () => {
+			dialogGlobalSettingsController["_viewModel"] = { isWhiteBackground: null } as any
+			settings.appSettings.isWhiteBackground = true
+
+			dialogGlobalSettingsController.updateSettingsFields(settings)
+
+			expect(dialogGlobalSettingsController["_viewModel"].isWhiteBackground).toBeTruthy()
+		})
+
+		it("should update viewModel.resetCameraIfNewFileIsLoaded with appSettings", () => {
+			dialogGlobalSettingsController["_viewModel"] = { resetCameraIfNewFileIsLoaded: null } as any
+			settings.appSettings.resetCameraIfNewFileIsLoaded = false
+
+			dialogGlobalSettingsController.updateSettingsFields(settings)
+
+			expect(dialogGlobalSettingsController["_viewModel"].resetCameraIfNewFileIsLoaded).toBeFalsy()
+		})
+	})
+
+	describe("applySettings", () => {
+		it("should call settingsService.updateSettings", () => {
+			settingsService.updateSettings = jest.fn()
+			const viewModel = { hideFlatBuildings: false, resetCameraIfNewFileIsLoaded: false }
+			dialogGlobalSettingsController["_viewModel"] = viewModel as any
```

Comment:
Instead of doing these things, add a function `setViewModel` to the test, that takes in one parameter for each field and sets it. This should shorten the code and remove the anys

Reply:
The viewModel is empty per default.

Change
```
function setEmptyViewModel() {
		dialogGlobalSettingsController["_viewModel"] = {
			hideFlatBuildings: null,
			maximizeDetailPanel: null,
			isWhiteBackground: null,
			resetCameraIfNewFileIsLoaded: null
		}
	}
```
to
```
function setViewModel(hideFlatBuildings: boolean, maximizeDetailPanel: boolean, isWhiteBackground: boolean, resetCameraIfNewFileIsLoaded: boolean) {
		dialogGlobalSettingsController["_viewModel"] = {
			hideFlatBuildings,
			maximizeDetailPanel,
			isWhiteBackground,
			resetCameraIfNewFileIsLoaded
		}
	}
```

so that this

```
it("should call settingsService.updateSettings", () => {
			settingsService.updateSettings = jest.fn()
			setEmptyViewModel()
			dialogGlobalSettingsController["_viewModel"].hideFlatBuildings = false
			dialogGlobalSettingsController["_viewModel"].resetCameraIfNewFileIsLoaded = false

			dialogGlobalSettingsController.applySettings()

			expect(settingsService.updateSettings).toHaveBeenCalledWith({ appSettings: dialogGlobalSettingsController["_viewModel"] })
		})
```

will become

```
it("should call settingsService.updateSettings", () => {
			settingsService.updateSettings = jest.fn()
			setViewModel(false, null, null, false)

			dialogGlobalSettingsController.applySettings()

			expect(settingsService.updateSettings).toHaveBeenCalledWith({ appSettings: dialogGlobalSettingsController["_viewModel"] })
		})
```

Another approach would be to make this function more abstract, so that the first parameter sets the value of the first key in the viewmodel. We could use that function in all controllers to setup the viewmodel in one line for a test.

Something I'm concerned about though:
It shortens code, if you need to setup more than one field of the viewModel, but if we only need one field of the viewModel for our test to work, it's probably easier to read when we write `myController.viewModel["key] = foo`
What do you think?

Reply:
I actually prefer it the other way around. Your approach might be less code, but it takes longer to understand and change the parameters if needed.

Reply:
Alright. But at least see my latest comment and remove the `setEmptyViewModel` function